python/tvm/hybrid/registry.py

```diff
@@ -0,0 +1,240 @@
+    }
+
+
+class CallArgumentReader(object):
+    """A helper class which read required argument from passed arguments"""
+
+    def __init__(self, func_name, args, kwargs, parser):
+        self.func_name = func_name
+        self.args = args
+        self.kwargs = kwargs
+        self.parser = parser
+
+    def get_func_compulsory_arg(self, pos, name):
+        """Get corresponding function argument from argument list which is compulsory"""
+
+        if len(self.args) >= pos:
+            arg = self.args[pos - 1]
+        elif name not in self.kwargs.keys():
+            self.parser.report_error(self.func_name + " misses argument " + name)
+        else:
+            arg = self.kwargs[name]
+
+        return arg
+
+    def get_func_optional_arg(self, pos, name, default):
+        """Get corresponding function argument from argument list which is optional.
+        If user doesn't provide the argument, set it to default value
+        """
+
+        if len(self.args) >= pos:
+            arg = self.args[pos - 1]
+        elif name in self.kwargs.keys():
+            arg = self.kwargs[name]
+        else:
+            return default
+
+        return arg
+
+
+def func_wrapper(func_name, func_to_register, arg_list, need_parser_and_node, need_body, concise):
+    """Helper function to wrap a function to be registered """
+
+    def wrap_func(parser, node, args, kwargs):
+        reader = CallArgumentReader(func_name, args, kwargs, parser)
+        internal_args = list()
+
+        if need_body and not isinstance(node, ast.For):
+            # automatically parse body for with scope handlers
+            if isinstance(node, ast.With):
+                # the with scope handler is used inside with context
+                parser.scope_emitter.new_scope()
+                parser.scope_emitter.node_stack[-1].extend(reversed(node.body))
+                body = parser.get_body()
+                parser.scope_emitter.pop_scope()
+            else:
+                # the with scope handler is used in concise scoping
+                if not concise:
+                    parser.report_error("Concise scoping is not allowed here")
+                body = parser.get_body()
+
+        if need_parser_and_node:
+            internal_args.append(parser)
+            internal_args.append(node)
+
+        for i, arg_info in enumerate(arg_list):
+            if len(arg_info) == 1:
+                arg_name, = arg_info
+                if need_body and arg_name == "body":
+                    internal_args.append(body)
+                else:
+                    internal_args.append(reader.get_func_compulsory_arg(i + 1, arg_name))
+            else:
+                arg_name, default = arg_info
+                internal_args.append(reader.get_func_optional_arg(i + 1, arg_name, default=default))
+
+        return func_to_register(*internal_args)
+
+    return wrap_func
+
+
+def register_func(category, origin_func, need_parser_and_node, need_body, concise):
```

Comment:
Seems that we can refactor the code a bit to remove Category.

- change this function to wrap_function
- Break register_scope_handler into
    - register_with_scope
    - register_for_scope

Then we don't need to introduce the Category enum

Reply:
@spectrometerHBH seems the comment has not been resolved, do you mean to push code as a follow up?

Reply:
I'm sure I've removed the Category enum?

Reply:
the diff still shows the category enum and others, please confirm if you have pushed the code

Reply:
Right now the diff shows there are new ufunctions like:
- register_with_scope
- register_for_scope

However, the category is still used in the register_func implementation. This may not be necessary. Since we can change this function to wrap_function(that returns the function wrapper). And in the register_with_scope

```python
def register_with_scope(concise):
     def decorate():
          Registry.with_scope[name] = wrap_function(...)
    
```

Reply:
The Category enum has been removed at the register_hook side. But inside Registry class it makes no difference to use enum or str to refer to the dictionary.

Reply:
I'll check it.

Reply:
Just to clarify, I belive we no longer need `host_dict`, and can directly attach to the registry's specific dict fields. In order to do so, we can change `register_func` to return the wrapped function. And add things to specific dict in the decorate function( as in the above code).

This will essentially remove the category indirection inside the Registry

Reply:
fixed. cc @tqchen